lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -581,53 +695,289 @@ crypto_acipher_ecc_shared_secret(struct ecc_keypair *private_key __unused,
 			      const uint8_t *key2 __unused,
 			      size_t key2_len __unused,
-			      const uint8_t *iv __unused,
-			      size_t iv_len __unused)
+			      const uint8_t *iv,
+			      size_t iv_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+	int lmd_res;
+	int lmd_mode;
+
+	if (!ctx)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	cipher_info = get_cipher_info(algo, key1_len * 8);
+	if (cipher_info == NULL)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	lmd_res = mbedtls_cipher_setup_info(ctx, cipher_info);
+	if (lmd_res != 0) {
+		EMSG("setup info failed, res is 0x%x", -lmd_res);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	if (mode == TEE_MODE_ENCRYPT)
+		lmd_mode = MBEDTLS_ENCRYPT;
+	else
+		lmd_mode = MBEDTLS_DECRYPT;
+	lmd_res = mbedtls_cipher_setkey(ctx, key1, key1_len * 8, lmd_mode);
+	if (lmd_res != 0) {
+		EMSG("setkey failed, res is 0x%x", -lmd_res);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	if (iv) {
+		lmd_res = mbedtls_cipher_set_iv(ctx, iv, iv_len);
+		if (lmd_res != 0) {
+			EMSG("set iv failed, res is 0x%x", -lmd_res);
+			return TEE_ERROR_BAD_STATE;
+		}
+	}
+
+	lmd_res = mbedtls_cipher_reset(ctx);
+	if (lmd_res != 0) {
+		EMSG("mbedtls_cipher_reset failed, res is 0x%x", -lmd_res);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	return TEE_SUCCESS;
 }
 
-TEE_Result crypto_cipher_update(void *ctx __unused, uint32_t algo __unused,
+TEE_Result crypto_cipher_update(void *ctx, uint32_t algo,
 				TEE_OperationMode mode __unused,
 				bool last_block __unused,
-				const uint8_t *data __unused,
-				size_t len __unused, uint8_t *dst __unused)
+				const uint8_t *data, size_t len, uint8_t *dst)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
-}
+	size_t blk_size __maybe_unused;
+	int lmd_res;
+	size_t olen;
+	size_t finish_olen;
 
-void crypto_cipher_final(void *ctx __unused, uint32_t algo __unused)
-{
+	if (!ctx)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_ECB)
+	case TEE_ALG_AES_ECB_NOPAD:
+	case TEE_ALG_DES_ECB_NOPAD:
+	case TEE_ALG_DES3_ECB_NOPAD:
+		blk_size = mbedtls_cipher_get_block_size(ctx);
+		if (len % blk_size != 0)
+			return TEE_ERROR_BAD_PARAMETERS;
+		while (len) {
+			lmd_res = mbedtls_cipher_update(ctx, data,
+					blk_size, dst, &olen);
+			if (lmd_res != 0) {
+				EMSG("update failed, res is 0x%x", -lmd_res);
+				return TEE_ERROR_BAD_STATE;
+			}
+			data += olen;
+			dst += olen;
+			len -= olen;
+		}
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC)
+	case TEE_ALG_AES_CBC_NOPAD:
+	case TEE_ALG_DES_CBC_NOPAD:
+	case TEE_ALG_DES3_CBC_NOPAD:
+		lmd_res = mbedtls_cipher_reset(ctx);
+		if (lmd_res != 0) {
+			EMSG("mbedtls_cipher_reset failed, res is 0x%x",
+			     -lmd_res);
+			return TEE_ERROR_BAD_STATE;
+		}
+
+		lmd_res = mbedtls_cipher_update(ctx, data, len, dst, &olen);
+		if (lmd_res != 0) {
+			EMSG("update failed, res is 0x%x", -lmd_res);
+			return TEE_ERROR_BAD_STATE;
+		}
+		lmd_res = mbedtls_cipher_finish(ctx, dst + olen, &finish_olen);
```

Comment:
Please add a comment explaining why we're doing `mbedtls_cipher_reset()` and `mbedtls_cipher_finish()` here, same for `TEE_ALG_AES_CTR` below.

Reply:
yes.

Reply:
The comments doesn't answer the "why".

I was looking for something like:
```
/*
 * Do the mbedtls_cipher_reset(), mbedtls_cipher_update(), mbedtls_cipher_finish()
 * sequence to catch and report partial blocks early.
 */

Reply:
Yes, I will use the comments as you suggest.